server/src/main/java/org/elasticsearch/index/fieldvisitor/IgnoredSourceFieldLoader.java

```diff
@@ -0,0 +1,142 @@
+                potentialFieldsInIgnoreSource.addAll(FallbackSyntheticSourceBlockLoader.splitIntoFieldPaths(fieldName));
+            }
+        }
+        this.potentialFieldsInIgnoreSource = potentialFieldsInIgnoreSource;
+    }
+
+    @Override
+    public LeafStoredFieldLoader getLoader(LeafReaderContext ctx, int[] docs) throws IOException {
+        var reader = sequentialReader(ctx);
+        var visitor = new SFV(potentialFieldsInIgnoreSource);
+        return new LeafStoredFieldLoader() {
+
+            private int doc = -1;
+
+            @Override
+            public void advanceTo(int doc) throws IOException {
+                if (doc != this.doc) {
+                    visitor.reset();
+                    reader.accept(doc, visitor);
+                    this.doc = doc;
+                }
+            }
+
+            @Override
+            public BytesReference source() {
+                return null;
+            }
+
+            @Override
+            public String id() {
+                return null;
+            }
+
+            @Override
+            public String routing() {
+                return null;
+            }
+
+            @Override
+            public Map<String, List<Object>> storedFields() {
+                return Map.of(IgnoredSourceFieldMapper.NAME, visitor.values);
+            }
+        };
+    }
+
+    @Override
+    public List<String> fieldsToLoad() {
+        return List.of(IgnoredSourceFieldMapper.NAME);
+    }
+
+    static class SFV extends StoredFieldVisitor {
+
+        boolean done;
+        final List<Object> values = new ArrayList<>();
+        final Set<String> potentialFieldsInIgnoreSource;
+
+        SFV(Set<String> potentialFieldsInIgnoreSource) {
+            this.potentialFieldsInIgnoreSource = potentialFieldsInIgnoreSource;
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            if (done) {
+                return Status.STOP;
+            } else if (IgnoredSourceFieldMapper.NAME.equals(fieldInfo.name)) {
+                return Status.YES;
+            } else {
+                return Status.NO;
+            }
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            var result = IgnoredSourceFieldMapper.decodeIfMatch(value, potentialFieldsInIgnoreSource);
+            if (result != null) {
+                // TODO: can't do this in case multiple entries for the same field name. (objects, arrays etc.)
+                // done = true;
+                values.add(result);
+            }
+        }
+
+        void reset() {
+            values.clear();
+            done = false;
+        }
+
+    }
+
+    static boolean supports(StoredFieldsSpec spec) {
+        return spec.requiresSource() == false
+            && spec.requiresMetadata() == false
+            && spec.requiredStoredFields().size() == 1
```

Comment:
When a query requests multiple values, the `ValuesFromManyReader#fieldsMoved()`, will merge the multiple `StoredFieldSpecs` into a single spec that specifies multiple fields. This check then fails (since it expects only 1 stored field in the spec), and we fall back to the default `StoredFieldLoader` implementation.